lib/posix.lua.in

```diff
@@ -519,27 +519,42 @@ if _DEBUG ~= false then
   end
 end
 
-local function prepare_args (pat_mark)
-  local pat, mark = "*", 0
-  if type (pat_mark) == "table" then
-    if type (pat_mark.pattern) == "string" then
-      pat = pat_mark.pattern
-    end
-    if pat_mark.MARK then
-      mark = M.GLOB_MARK
-    end
-  elseif type (pat_mark) == "string" then
-    pat = pat_mark
-  end
-
-  return pat, mark
+--- Find all files in this directory matching a shell pattern.
+-- The flag MARK appends a trailing slash to matches that are directories.
+-- @function glob
+-- @param[opt={}] string|table args either a glob pattern string or a table
+--    with optional keys `pattern` and `MARK`. If no argument is given, the
+--    function uses a default pattern of "*" and MARK is false.
+-- @treturn A table of zero or more matches to the pattern.
+local function glob (args)
+  local posix_glob = require "posix.glob"
+  -- Support previous `glob ".*"` style calls.
+  if type(args) == "string" then
+    args = {pattern=args}
+  elseif type(args) ~= "table" then
+    args = {}
+  end
+  local flags = 0
+  if args.MARK then
+    flags = posix_glob.GLOB_MARK
+  end
+  return posix_glob.glob(args.pattern, flags)
 end
 
-local cglob = M.glob
-
-M.glob = function (...)
-  local pat, mark = prepare_args(...)
-  return cglob(pat, mark)
+if _DEBUG ~= false then
+  M.glob = function (...)
+    local argt = {...}
+    local argtype = type(argt[1])
+    local validtypes = { ["table"] = true, ["string"] = true, ["nil"] = true }
+    if validtypes[argtype] ~= true then
+      argtypeerror ("glob", 1, "table, string, or nil", argt[1])
```

Comment:
I'm not sure that's quite the right diagnosis (though I get you what you mean about not needing to mention `nil`).

Here's what I had to do to squelch the errors locally:

      M.glob = function (...)
        local argt = {...}
        local argtype = type(argt[1])
        if validtypes[argtype] ~= true then
          argtypeerror ("?", 1, "table, string or nil", argt[1]) -- "fn" works too
        elseif #argt > 1 then
          toomanyargerror ("?", 1, #argt) -- "fn" works too
        end
        return glob (...)
      end

1. Remove the Oxford comma (as you noted), but keep the `nil` in the list.
2. Change the function name to either `?` or `fn`, presumably because specl knows that it's an anonymous function. If I keep the name as `"glob"`.

Without the second change, this is what I get:

    - posix file system glob with bad arguments diagnoses argument #1 type not table, string or nil:
      FAILED expectation 1: expecting an error containing any of:
      | bad argument #1 to '?' (table, string or nil expected, got boolean)
      or:
      | bad argument #1 to 'fn' (table, string or nil expected, got boolean)
      but got:
      | ./lib/posix/init.lua:552: bad argument #1 to 'glob' (table, string or nil expected, got boolean)
    
    - posix file system glob with bad arguments diagnoses more than maximum of 1 arguments:
      FAILED expectation 1: expecting an error containing any of:
      | bad argument #2 to 'fn' (no more than 1 argument expected, got 2)
      or:
      | bad argument #2 to '?' (no more than 1 argument expected, got 2)
      but got:
      | ./lib/posix/init.lua:554: bad argument #2 to 'glob' (no more than 1 argument expected, got 2)

For the moment, I've removed the serial comma: that's easy enough. (I don't even have strong feelings about it! I think I do it by habit, frankly.) But I'm leaving the name of the function as `"glob"` since I'm guessing that's a bug in specl (or that I'm doing something larger wrong).